tokio/src/runtime/blocking/pool.rs

```diff
@@ -317,6 +317,8 @@ impl Spawner {
         R: Send + 'static,
     {
         let fut = BlockingTask::new(func);
+        #[cfg(feature = "test-util")]
+        let fut = crate::time::inhibit_auto_advance(fut);
```

Comment:
I think this has a race condition. The JoinHandle for a task is notified _after_ the destructor of the future runs, so this will _first_ re-enable auto-advance, then notify the task. If the runtime notices that auto-advance has been reenabled, but has not yet received the notification on the `JoinHandle`, then it might advance the timer even though it should just have polled the task waiting for the `JoinHandle`.

A loom test would probably catch it.

Reply:
You might want to join our discord server to discuss this issue in more details.

Reply:
Thanks. I think I see the problem. I'll say hi in discord, probably Tuesday afternoon Nashville time (US Central, UTC-5).

Reply:
I'm in Europe, so I am unlikely to be around in your afternoon. That said, we have some US maintainers as well, so there should be someone else around.

Reply:
OK, I posted a Loom test that fails because of this bug.

How should I fix this? It's easy to decorate the future that's sent to the blocking-task-runner thread, and thus transfer the `AutoAdvanceInhibit` to the thread. It's not so easy to change the behavior of `<JoinHandle<T> as Future>::poll` to bring the inhibit-value back to the tokio thread and drop it there. I think that's necessary to close this gap.

Given that the same `JoinHandle` type is used for all tasks, I don't see how to fix it without adding a bit (encoding an `Option<AutoAdvanceInhibit>`) to every task. Maybe it could be stored in the vtable somehow instead? Two different implementations of `try_read_output`? Every option looks invasive.

Reply:
The `release` method in `src/runtime/blocking/schedule.rs` is called after waking the waker. It should be possible to add a field to `NoopSchedule` that keeps track of whether auto-advance should be re-enabled on release, and have it do so from the schedule call.

(It should probably be renamed from `NoopSchedule` to something else though.)